From 7d184c43d34008fabe79e9dc7cdea5593fc9116a Mon Sep 17 00:00:00 2001
From: Prarit Bhargava <prarit@redhat.com>
Date: Wed, 6 Aug 2014 14:33:25 -0400
Subject: [acpi] processor: Fix STARTING/DYING action in acpi_cpu_soft_notify()

Message-id: <1407335605-19425-1-git-send-email-prarit@redhat.com>
Patchwork-id: 87505
O-Subject: [RHEL7.1 PATCH BZ 1099594] ACPI / processor: Fix STARTING/DYING action in acpi_cpu_soft_notify()
Bugzilla: 1099594
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Dean Nelson <dnelson@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1099594

Date: Wed Aug  6 10:18:29 EDT 2014
Build OS: Red Hat Enterprise Linux Server release 7.0 (Maipo)
System name: intel-canoepass-05.lab.bos.redhat.com with -j24
Built on: kernel-3.10.0-142.el7
Arch built: ppc64 s390x x86_64

commit 8da8373447d6a57a5a9f55233d35beb15d92d0d2
Author: Toshi Kani <toshi.kani@hp.com>
Date:   Thu May 8 07:58:59 2014 -0600

    ACPI / processor: Fix STARTING/DYING action in acpi_cpu_soft_notify()

    During CPU online/offline testing on a large system, one of the
    processors got stuck after the message "bad: scheduling from the
    idle thread!".  The problem is that acpi_cpu_soft_notify() calls
    acpi_bus_get_device() for all action types.  CPU_STARTING and
    CPU_DYING do not allow the notify handlers to sleep.  However,
    acpi_bus_get_device() can sleep in acpi_ut_acquire_mutex().

    Change acpi_cpu_soft_notify() to return immediately for CPU_STARTING
    and CPU_DYING as they have no action in this handler.

    Signed-off-by: Toshi Kani <toshi.kani@hp.com>
    Signed-off-by: Rafael J. Wysocki <rafael.j.wysocki@intel.com>

We don't have any systems that support physical hotplug so I did some
pretty extensive testing with softplug.  I didn't see any issues.
Successfully tested by me.

Cc: Lenny Szubowicz <lszubowi@redhat.com>
Cc: Tony Camuso <tcamuso@redhat.com>
Cc: Dean Nelson <dnelson@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/acpi/processor_driver.c b/drivers/acpi/processor_driver.c
index 870eaf5..aa12472 100644
--- a/drivers/acpi/processor_driver.c
+++ b/drivers/acpi/processor_driver.c
@@ -127,6 +127,13 @@ static int acpi_cpu_soft_notify(struct notifier_block *nfb,
  struct acpi_processor *pr = per_cpu(processors, cpu);
  struct acpi_device *device;
 
+ /*
+  * CPU_STARTING and CPU_DYING must not sleep. Return here since
+  * acpi_bus_get_device() may sleep.
+  */
+ if (action == CPU_STARTING || action == CPU_DYING)
+  return NOTIFY_DONE;
+
  if (!pr || acpi_bus_get_device(pr->handle, &device))
   return NOTIFY_DONE;
 
-- 
1.7.1